kimchi/src/plonk_sponge.rs

```diff
@@ -59,36 +59,25 @@ impl<Fr: PrimeField> FrSponge<Fr> for DefaultFrSponge<Fr, SC> {
     fn absorb_evaluations<const N: usize>(&mut self, e: [&ProofEvaluations<Vec<Fr>>; N]) {
         self.last_squeezed = vec![];
 
-        let e = ProofEvaluations::transpose(e);
-
-        let mut points = vec![
-            &e.z,
-            &e.generic_selector,
-            &e.poseidon_selector,
-            &e.w[0],
-            &e.w[1],
-            &e.w[2],
-            &e.w[3],
-            &e.w[4],
-            &e.w[5],
-            &e.w[6],
-            &e.w[7],
-            &e.w[8],
-            &e.w[9],
-            &e.w[10],
-            &e.w[11],
-            &e.w[12],
-            &e.w[13],
-            &e.w[14],
-            &e.s[0],
-            &e.s[1],
-            &e.s[2],
-            &e.s[3],
-            &e.s[4],
-            &e.s[5],
-        ];
-
-        if let Some(l) = e.lookup.as_ref() {
+        let ProofEvaluations {
+            w,
+            z,
+            s,
+            lookup,
+            generic_selector,
+            poseidon_selector,
+        } = ProofEvaluations::transpose(e);
```

Comment:
I think it's a better approach to use the code here https://github.com/o1-labs/proof-systems/issues/221#issuecomment-1246346523 so that we can remove this `transpose` function (which is error prone and not really elegant imo)

Reply:
btw we don't have to create an `Absorbable` trait, we can just put the logic here in this function:

```
let [zeta_evals, zeta_omega_evals] = &evals;

        // we interleave points from each evaluations,
        // it makes it easier to absorb them in the verifier circuit
        let points: Vec<F> = zeta_evals
            .iter()
            .interleave(zeta_omega_evals.iter())
            .collect();

        sponge.absorb_multiple(&points);
```

Reply:
@mimoo 
I think this also needs the `ChunkEvaluations` implementation first.
The reason is that I think there is mismatch between the `ProofEvaluations` we think.
For example, the `ProofEvaluations` I mean is like
```

// TODO: this should really be vectors here, perhaps create another type for chunked evaluations?
/// Polynomial evaluations contained in a `ProverProof`.
/// - **Chunked evaluations** `Field` is instantiated with vectors with a length that equals the length of the chunk
/// - **Non chunked evaluations** `Field` is instantiated with a field, so they are single-sized#[serde_as]
#[serde_as]
#[derive(Clone, Serialize, Deserialize)]
#[serde(bound(
    serialize = "Vec<o1_utils::serialization::SerdeAs>: serde_with::SerializeAs<Field>",
    deserialize = "Vec<o1_utils::serialization::SerdeAs>: serde_with::DeserializeAs<'de, Field>"
))]
pub struct ProofEvaluations<Field> {
    /// witness polynomials
    #[serde_as(as = "[Vec<o1_utils::serialization::SerdeAs>; COLUMNS]")]
    pub w: [Field; COLUMNS],
    /// permutation polynomial
    #[serde_as(as = "Vec<o1_utils::serialization::SerdeAs>")]
    pub z: Field,
    /// permutation polynomials
    /// (PERMUTS-1 evaluations because the last permutation is only used in commitment form)
    #[serde_as(as = "[Vec<o1_utils::serialization::SerdeAs>; PERMUTS - 1]")]
    pub s: [Field; PERMUTS - 1],
    /// lookup-related evaluations
    pub lookup: Option<LookupEvaluations<Field>>,
    /// evaluation of the generic selector polynomial
    #[serde_as(as = "Vec<o1_utils::serialization::SerdeAs>")]
    pub generic_selector: Field,
    /// evaluation of the poseidon selector polynomial
    #[serde_as(as = "Vec<o1_utils::serialization::SerdeAs>")]
    pub poseidon_selector: Field,
}
```

While yours is like
```
/// Polynomial evaluations contained in a `ProverProof`.
#[serde_as]
#[derive(Clone, Serialize, Deserialize)]
pub struct ProofEvaluations<F>
where
    F: Field,
{
    /// witness polynomials
    #[serde_as(as = "[o1_utils::serialization::SerdeAs; COLUMNS]")]
    pub w: [F; COLUMNS],

    /// permutation polynomial
    #[serde_as(as = "o1_utils::serialization::SerdeAs")]
    pub z: F,

    /// permutation polynomials
    /// (PERMUTS-1 evaluations because the last permutation is only used in commitment form)
    #[serde_as(as = "[o1_utils::serialization::SerdeAs; PERMUTS - 1]")]
    pub s: [F; PERMUTS - 1],

    /// lookup-related evaluations
    #[serde(bound = "LookupEvaluations<F>: Serialize + DeserializeOwned")]
    pub lookup: Option<LookupEvaluations<F>>,

    /// evaluation of the generic selector polynomial
    #[serde_as(as = "o1_utils::serialization::SerdeAs")]
    pub generic_selector: F,

    /// evaluation of the poseidon selector polynomial
    #[serde_as(as = "o1_utils::serialization::SerdeAs")]
    pub poseidon_selector: F,
}
```
(https://github.com/o1-labs/proof-systems/blob/501616da46b26d71114f92dc4524dbf27ee97a2d/kimchi/src/proof.rs#L44-L75)
Simply, I think the `transpose` is needed current `ProofEvaluations`, and we can introduce `iter` when the `ProofEvaluations` is updated as yours(I mean `ChunkEvaluations` introduced).
Right?

Reply:
perhaps it would be better to have a `ChunkedEvaluation` type before that change to make it simpler though :o

Reply:
Yes, I also think that way, @mimoo .
Only when we introduce the `ChunkedEvaluations` type to the system, I can tackle the 2 issues I am working on, in best way.

Reply:
One request, @mimoo 
Can you make time to create the PR of `ChunkedEvaluations` type? or at least the prototype that I can reference and improve?
The reason is that I am still not much familiar with whole system.
(Even though I have been reading several books(mina book, halo2 book, real-world cryptography ...), the implementation needs some more practice for me. 😄 )

Reply:
I'll try to add some more information on https://github.com/o1-labs/proof-systems/issues/759

Reply:
Thanks so much for detailed guide! @mimoo 
I will try to take more look at the system & create the PR myself.